compose/utils.py

```diff
@@ -83,6 +86,71 @@ def get_output_stream(stream):
     return codecs.getwriter('utf-8')(stream)
 
 
+def stream_as_text(stream):
+    """Given a stream of bytes or text, if any of the items in the stream
+    are bytes convert them to text.
+
+    This function can be removed once docker-py returns text streams instead
+    of byte streams.
+    """
+    for data in stream:
+        if not isinstance(data, six.text_type):
+            data = data.decode('utf-8')
+        yield data
+
+
+def line_splitter(buffer, separator=u'\n'):
+    index = buffer.find(six.text_type(separator))
+    if index == -1:
+        return None, None
+    return buffer[:index + 1], buffer[index + 1:]
+
+
+def split_buffer(stream, splitter=None, decoder=lambda a: a):
+    """Given a generator which yields strings and a splitter function,
+    joins all input, splits on the separator and yields each chunk.
+
+    Unlike string.split(), each chunk includes the trailing
+    separator, except for the last one if none was found on the end
+    of the input.
+    """
+    splitter = splitter or line_splitter
+    buffered = six.text_type('')
+
+    for data in stream_as_text(stream):
+        buffered += data
+        while True:
+            item, rest = splitter(buffered)
+            if not item:
+                break
+
+            buffered = rest
+            yield item
+
+    if buffered:
+        yield decoder(buffered)
```

Comment:
I don't think so. That would result in different types being returned from the stream.

Most of the items that get returned are from the `yield item` above, and they're all `dict` type (already loaded json object).  I had to add the decoder because we should get the same response type from the last yield. Without it we'd get a string for the last item. I can't re-use the same splitter function because in the newline splitter version it needs to work without a trailing newline.